File: fastApiProject/service/guessWordService.py
from sqlalchemy.orm import Session, defer
from sqlalchemy import func
from typing import List, Optional
from datetime import datetime
import json

from entity.GuessWord import GuessWord
from database import GuessWordModel
from service.embeddingService import embedding_service
from service.redisService import redis_service


class GuessWordService:
    def __init__(self):
        pass

    def create_guess_word(self, db: Session, word: str, hint: Optional[str] = None,
                          difficulty: int = 1) -> GuessWord:
        """创建目标字"""
        db_word = GuessWordModel(
            word=word,
            hint=hint,
            difficulty=difficulty,
            is_passed=False,
            pass_count=0,
            created_at=datetime.now()
        )
        db.add(db_word)
        db.commit()
        db.refresh(db_word)

        # 生成并保存向量
        embedding_vec = embedding_service.encode_single(word)
        db_word.embedding = embedding_vec
        db.commit()
        db.refresh(db_word)

        # 使缓存失效
        redis_service.invalidate_guess_words_cache()

        return self._to_pydantic(db_word)

    def get_guess_word_by_id(self, db: Session, word_id: int) -> Optional[GuessWord]:
        """根据ID获取目标字"""
        db_word = db.query(GuessWordModel).filter(
            GuessWordModel.id == word_id
        ).first()

        if not db_word:
            return None

        return self._to_pydantic(db_word)

    def list_guess_words(self, db: Session, skip: int = 0, limit: int = 10) -> List[GuessWord]:
        """列出目标字列表（使用Redis缓存）"""
        # 尝试从Redis获取缓存
        cached_data = redis_service.get_cached_guess_words_list()
        if cached_data:
            try:
                all_words = json.loads(cached_data)
                paginated = all_words[skip:skip + limit]
                for i in paginated:
                    i.word = str(len(i.word))
                    i.embedding = []
                return [GuessWord(**w) for w in paginated]
            except Exception:
                pass

        # 缓存未命中，从数据库查询（排除embedding字段以提升性能）
        db_words = db.query(GuessWordModel).options(
            defer(GuessWordModel.embedding)
        ).order_by(
            GuessWordModel.created_at.desc()
        ).all()
        res = [self._to_pydantic(w, include_embedding=False) for w in db_words]
        for i in res:
            i.word = str(len(i.word))
            i.embedding = []

        # 缓存结果
        cache_data = [self._to_pydantic(w, include_embedding=False).model_dump(mode='json') for w in db_words]
        redis_service.cache_guess_words_list(json.dumps(cache_data, default=str))

        return res[skip:skip + limit]

    def mark_as_passed(self, db: Session, word_id: int) -> Optional[GuessWord]:
        """标记目标字为已通过"""
        db_word = db.query(GuessWordModel).filter(
            GuessWordModel.id == word_id
        ).first()

        if not db_word:
            return None

        db_word.is_passed = True
        db_word.pass_count = db_word.pass_count + 1
        db.commit()
        db.refresh(db_word)

        # 使缓存失效
        redis_service.invalidate_guess_words_cache()

        return self._to_pydantic(db_word)

    def get_total_count(self, db: Session) -> int:
        """获取目标字总数"""
        return db.query(func.count(GuessWordModel.id)).scalar()

    def get_passed_count(self, db: Session) -> int:
        """获取已通过的目标字数量"""
        return db.query(GuessWordModel).filter(GuessWordModel.is_passed == True).count()

    def _to_pydantic(self, db_word: GuessWordModel, include_embedding: bool = True) -> GuessWord:
        """ORM模型转Pydantic实体"""
        return GuessWord(
            id=db_word.id,
            word=db_word.word,
            hint=db_word.hint,
            difficulty=db_word.difficulty,
            is_passed=db_word.is_passed,
            pass_count=db_word.pass_count,
            created_at=db_word.created_at,
            embedding=db_word.embedding if include_embedding else []
        )


# 全局单例实例
guess_word_service = GuessWordService()

File: blog-agent/agent/src/main/java/com/xiafan/agent/service/GuessWordService.java
package com.xiafan.agent.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.xiafan.agent.config.AppProperties;
import com.xiafan.agent.entity.GuessWord;
import com.xiafan.agent.repository.GuessWordRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/** Mirrors guessWordService.py (Redis-cached list, embedding on create, pass marking). */
@Service
public class GuessWordService {

    private static final Logger log = LoggerFactory.getLogger(GuessWordService.class);

    private final GuessWordRepository repository;
    private final EmbeddingService embeddingService;
    private final RedisService redisService;
    private final AppProperties props;
    private final ObjectMapper objectMapper;

    public GuessWordService(GuessWordRepository repository, EmbeddingService embeddingService,
                            RedisService redisService, AppProperties props, ObjectMapper objectMapper) {
        this.repository = repository;
        this.embeddingService = embeddingService;
        this.redisService = redisService;
        this.props = props;
        this.objectMapper = objectMapper;
    }

    public GuessWord createGuessWord(String word, String hint, int difficulty) {
        GuessWord created = repository.insert(word, hint, difficulty);
        try {
            List<Double> embedding = embeddingService.encodeSingle(word);
            repository.updateEmbedding(created.getId(), embedding);
        } catch (Exception e) {
            log.warn("embedding generation failed for guess word {}: {}", word, e.getMessage());
        }
        redisService.invalidateGuessWordsCache();
        return created;
    }

    public Optional<GuessWord> getGuessWordById(int wordId) {
        return repository.findById(wordId);
    }

    /** Paginated list with Redis caching; word is masked as its length and embedding omitted (mirrors Python). */
    public List<GuessWord> listGuessWords(int skip, int limit) {
        Optional<String> cached = redisService.getCachedGuessWordsList();
        if (cached.isPresent()) {
            try {
                List<Map<String, Object>> allWords = objectMapper.readValue(cached.get(),
                        new com.fasterxml.jackson.core.type.TypeReference<List<Map<String, Object>>>() {
                        });
                List<Map<String, Object>> paginated = paginate(allWords, skip, limit);
                List<GuessWord> result = new ArrayList<>();
                for (Map<String, Object> w : paginated) {
                    result.add(toMaskedEntity(w));
                }
                return result;
            } catch (Exception e) {
                log.warn("guess-words cache parse failed, falling back to DB: {}", e.getMessage());
            }
        }
        List<GuessWord> dbWords = repository.findAll();
        List<Map<String, Object>> cacheData = new ArrayList<>();
        for (GuessWord w : dbWords) {
            cacheData.add(toCacheEntry(w));
        }
        try {
            redisService.cacheGuessWordsList(objectMapper.writeValueAsString(cacheData),
                    props.getGuess().getListCacheTtl());
        } catch (Exception e) {
            log.warn("guess-words cache write failed: {}", e.getMessage());
        }
        if (dbWords.size() > 0) {
            List<Map<String, Object>> paginated = paginate(cacheData, skip, limit);
            List<GuessWord> result = new ArrayList<>();
            for (Map<String, Object> w : paginated) {
                result.add(toMaskedEntity(w));
            }
            return result;
        }
        return List.of();
    }

    public Optional<GuessWord> markAsPassed(int wordId) {
        if (repository.markAsPassed(wordId) == 0) {
            return Optional.empty();
        }
        redisService.invalidateGuessWordsCache();
        return repository.findById(wordId);
    }

    public int getTotalCount() {
        return repository.countAll();
    }

    public int getPassedCount() {
        return repository.countPassed();
    }

    public int similarityCacheTtl() {
        return props.getGuess().getSimilarityCacheTtl();
    }

    private Map<String, Object> toCacheEntry(GuessWord w) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", w.getId());
        m.put("word", w.getWord());
        m.put("hint", w.getHint());
        m.put("difficulty", w.getDifficulty());
        m.put("is_passed", w.isPassed());
        m.put("pass_count", w.getPassCount());
        m.put("embedding", List.of());
        m.put("created_at", String.valueOf(w.getCreatedAt()));
        return m;
    }

    private static List<Map<String, Object>> paginate(List<Map<String, Object>> all, int skip, int limit) {
        int from = Math.min(skip, all.size());
        int to = Math.min(skip + limit, all.size());
        return all.subList(from, to);
    }

    private static GuessWord toMaskedEntity(Map<String, Object> w) {
        Object word = w.get("word");
        String masked = String.valueOf(word == null ? "" : word.toString().length());
        GuessWord e = new GuessWord();
        e.setId(((Number) w.getOrDefault("id", 0)).intValue());
        e.setWord(masked);
        e.setHint((String) w.get("hint"));
        e.setDifficulty(((Number) w.getOrDefault("difficulty", 1)).intValue());
        e.setPassed(Boolean.TRUE.equals(w.get("is_passed")));
        e.setPassCount(((Number) w.getOrDefault("pass_count", 0)).intValue());
        e.setEmbedding(List.of());
        return e;
    }
}